acme/dns_challenge.go

```diff
@@ -60,50 +61,129 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 		}
 	}()
 
-	fqdn, _, _ := DNS01Record(domain, keyAuth)
+	fqdn, value, _ := DNS01Record(domain, keyAuth)
 
-	preCheckDNS(domain, fqdn)
+	err = waitFor(60, 2, func() (bool, error) {
+		return preCheckDNS(domain, fqdn, value)
+	})
 
 	return s.validate(s.jws, domain, chlng.URI, challenge{Resource: "challenge", Type: chlng.Type, Token: chlng.Token, KeyAuthorization: keyAuth})
 }
 
-func checkDNS(domain, fqdn string) bool {
-	// check if the expected DNS entry was created. If not wait for some time and try again.
+// checkDnsPropagation queries all nameservers that are authoritative for 'domain'
+// to check whether DNS propagation for the expected DNS record is complete.
+func checkDnsPropagation(domain, fqdn, value string) (bool, error) {
+	nameservers, err := lookupNameservers(toFqdn(domain))
```

Comment:
>  ns1.bar.com could have a record _acme-challenge.foo.bar.com IN NS ns.confusion.party – delegating that specific name to a different nameserver, in which case this function should query ns.confusion.party instead.

No, i don't think it should. Because `ns.confusion.party` is **not** the authority for the certificate domain ` foo.bar.com`.

The purpose of the ACME challenge is to prove ownership of the domain for which a certificate is requested. In your scenario you don't show authority for the certificate domain `foo.bar.com` because you did not create a record in the `bar.com` zone. `ns1.bar.com` might have delegated `_acme-challenge.foo.bar.com` to you, but that does not mean that you have authority over `bar.com`.